### PR TITLE
Disable web editor finalized contest and other features

### PR DESCRIPTION
**Pull request by zapotocnylubos**
Solves

1. https://github.com/delta-cs/seminar-domjudge/issues/115
2. https://github.com/delta-cs/seminar-domjudge/issues/62
3. https://github.com/delta-cs/seminar-domjudge/issues/46
4. partially https://github.com/delta-cs/seminar-domjudge/issues/117
5. improves https://github.com/delta-cs/seminar-domjudge/issues/30
6. add check, if code in editor has changes
7. on problemset page, changes link from "test" to "Problem"

### CODE REVIEW

**Comment by zapotocnylubos** [comment]
Snímek obrazovky 2024-03-18 v 18 30 34
Snímek obrazovky 2024-03-18 v 18 20 42
Snímek obrazovky 2024-03-18 v 18 05 26
Snímek obrazovky 2024-03-18 v 17 49 12
Snímek obrazovky 2024-03-18 v 13 25 14
Snímek obrazovky 2024-03-18 v 13 19 53
Snímek obrazovky 2024-03-18 v 13 19 47
Snímek obrazovky 2024-03-18 v 13 19 33
Snímek obrazovky 2024-03-18 v 13 19 22
Snímek obrazovky 2024-03-18 v 13 19 10

File: webapp/src/Controller/Team/EditorController.php [hunk 1]
```diff
@@ -106,6 +106,16 @@ public function viewAction(Request $request, int $probId, string $langId): Respo
         $team = $this->dj->getUser()->getTeam();
         $contest = $this->dj->getCurrentContest($team->getTeamid());
 
+        $contestProblemsIds = [];
+        foreach ($contest->getProblems() as $contestProblem) {
+            $contestProblemsIds[] = $contestProblem->getProbid();
+        }
+
+        if (!in_array($problem->getProbid(), $contestProblemsIds)) {
```
Comment (matous-volf) [inline webapp/src/Controller/Team/EditorController.php]: neslo by tu toto?

```php
if (!$contest->getProblems()->map(fn (Problem $p) => $p->getProbid())->contains($problem->getProbid()))
```

File: webapp/src/Controller/Team/EditorController.php [hunk 2]
```diff
@@ -106,6 +106,16 @@ public function viewAction(Request $request, int $probId, string $langId): Respo
...
+            $this->addFlash('warning', 'This problem is not part of the current contest');
```
Comment (matous-volf) [inline webapp/src/Controller/Team/EditorController.php:114]: tohle znamena, ze uzivatel musi pred otevrenim editoru switchnout na contest, ve kterem problem dany je?

File: webapp/src/Controller/Team/EditorController.php (hunk 1 above)
Comment (zapotocnylubos) [reply webapp/src/Controller/Team/EditorController.php]: jj, diky, upraveno

File: webapp/src/Controller/Team/EditorController.php (hunk 2 above)
Comment (zapotocnylubos) [reply webapp/src/Controller/Team/EditorController.php:114]: upravil jsem i to generovani linku, takze
1. generuje se link s cid https://github.com/delta-cs/seminar-domjudge/blob/cf126830080b1326331be743976255db39703dc9/webapp/templates/partials/problem_list.html.twig#L142
2. pridal jsem tuto routu do cid-aware middleware https://github.com/delta-cs/seminar-domjudge/blob/cf126830080b1326331be743976255db39703dc9/webapp/config/services.yaml#L22

Comment (matous-volf) [reply webapp/src/Controller/Team/EditorController.php:114]: takze ten warning je pro pripady, kdy si uzivatel prepise `cid` v url? nebo jak jinak by se mohl dostat do editoru problemu, ktery neni v current contestu?

Comment (zapotocnylubos) [reply webapp/src/Controller/Team/EditorController.php:114]: muzes si editovat a pri tom pouzit contest switcher

Comment (matous-volf) [reply webapp/src/Controller/Team/EditorController.php:114]: ajo

File: webapp/templates/team/team_editor.html.twig [hunk 3]
```diff
@@ -31,6 +31,40 @@
...
+                    <ul class="list-unstyled ml-4 mt-1 mb-1">
+                        {% for contest_problem in problem_available_contests %}
+                            {% set contest = contest_problem.contest %}
+                            <li>
+                                <a href="{{ path('team_problems', {'cid': contest.cid}) }}">
+                                    <i class="fas fa-book-open"></i> {{ contest.name }}
+                                </a>
+                            </li>
+                        {% endfor %}
+                    </ul>
```
Comment (matous-volf) [inline webapp/templates/team/team_editor.html.twig:56]: tohle je hodne nice. jeste detail, jestli nechceme misto knizky pouzit trophy ikonku?

Comment (zapotocnylubos) [reply webapp/templates/team/team_editor.html.twig:56]: ono te to bere primo na ten problemset, ktery pouziva vsude knizku, tak jsem se drzel toho

Comment (matous-volf) [reply webapp/templates/team/team_editor.html.twig:56]: joo, jasne

**Review by matous-volf — APPROVED** [review]

**Pull request merged**